src/ln/msgs.rs

```diff
@@ -382,9 +384,94 @@ impl NetAddress {
 			&NetAddress::OnionV3 {..} => { 4 },
 		}
 	}
+
+	/// Strict byte-length of address descriptor, 1-byte type not recorded
+	fn len(&self) -> u16 {
+		match self {
+			&NetAddress::IPv4 { .. } => { 6 },
+			&NetAddress::IPv6 { .. } => { 18 },
+			&NetAddress::OnionV2 { .. } => { 12 },
+			&NetAddress::OnionV3 { .. } => { 37 },
+		}
+	}
 }
 
-#[derive(Clone)]
+impl Writeable for NetAddress {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		match self {
+			&NetAddress::IPv4 { ref addr, ref port } => {
+				1u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::IPv6 { ref addr, ref port } => {
+				2u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV2 { ref addr, ref port } => {
+				3u8.write(writer)?;
+				addr.write(writer)?;
+				port.write(writer)?;
+			},
+			&NetAddress::OnionV3 { ref ed25519_pubkey, ref checksum, ref version, ref port } => {
+				4u8.write(writer)?;
+				ed25519_pubkey.write(writer)?;
+				checksum.write(writer)?;
+				version.write(writer)?;
+				port.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl<R: ::std::io::Read>  Readable<R> for NetAddress {
+	fn read(reader: &mut R) -> Result<NetAddress, DecodeError> {
+		let byte = <u8 as Readable<R>>::read(reader)?;
+		match byte {
+			1 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::IPv4 {
+					addr,
+					port,
+				})
+			},
+			2 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::IPv6 {
+					addr,
+					port,
+				})
+			},
+			3 => {
+				let addr = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::OnionV2 {
+					addr,
+					port,
+				})
+			},
+			4 => {
+				let ed25519_pubkey = Readable::read(reader)?;
+				let checksum = Readable::read(reader)?;
+				let version = Readable::read(reader)?;
+				let port = Readable::read(reader)?;
+				Ok(NetAddress::OnionV3 {
+					ed25519_pubkey,
+					checksum,
+					version,
+					port,
+				})
+			}
+			_ => return Err(DecodeError::InvalidValue { byte: Some(byte) }),
```

Comment:
Hmm, now that I think about it more I'm rather unconvinced of this approach. If, eg, at some point we check the ed25519_pubkey is valid in OnionV3 or the ports are non-0 we may want to return an InvalidValue but not have extra bytes. Maybe it makes sense to return a Result<enum { Ok(NetAddress), UnknownType(u8) }, DecodeError>? This would also make this patch much smaller, which is nice.

Reply:
Added a NetAddess::Unknown { descriptor }, cleaner than an enum, Agree, cleaner and better do not touch InvalidValue so let it free to be return for an invalid ed25519_pubkey, zero ports or constrained fields in future addresses.